src/Conversion/ONNXToTOSA/NN/MaxPoolSingleOut.cpp

```diff
@@ -0,0 +1,160 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//===---------------- MaxPoolSingleOut.cpp - MaxPoolSingleOut Op-----------===//
+//
+// Copyright (c) 2022 Advanced Micro Devices, Inc.
+//
+// =============================================================================
+//
+// This file lowers ONNX MaxpoolSingleOut operator to TOSA dialect.
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Dialect/Tosa/IR/TosaOps.h"
+#include "mlir/IR/Attributes.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "src/Conversion/ONNXToTOSA/DialectBuilder.hpp"
+#include "src/Conversion/ONNXToTOSA/ONNXToTOSACommon.hpp"
+#include "src/Conversion/ONNXToTOSA/ONNXToTOSALegalizeUtils.hpp"
+#include "src/Dialect/ONNX/ONNXOps/ShapeHelper.hpp"
+#include "llvm/ADT/ArrayRef.h"
+#include <cstdint>
+#include <src/Dialect/Mlir/IndexExpr.hpp>
+
+using namespace mlir;
+
+namespace onnx_mlir {
+
+namespace {
+
+// This calculates the values that need to be added to the padding in order to
+// simulate the ceil mode
+llvm::SmallVector<int64_t> getCeilConstants(llvm::ArrayRef<int64_t> inputShape,
+    const ONNXMaxPoolSingleOutOpShapeHelper &shapeHelper, int64_t ceilMode) {
+  // This avoids having more if statements when creating the padding const.
+  if (ceilMode == 0)
+    return llvm::SmallVector<int64_t>{0, 0};
+
+  SmallVector<int64_t, 4> kernelShapeVec =
+      tosa::createInt64VectorFromIndexExpr(shapeHelper.kernelShape);
```

Comment:
Perfect usage of ShapeHelper, where you reuse work done in shape inference to extract all you need without further massaging of parameters.

I see that `tosa::createInt64VectorFromIndexExpr` is not a functionality that already exists in IndexExpr. Here is what is supported:, extract a shape (using -1 for dynamic) as well as others.
```C++
class IndexExpr {
...
  // Helpers for list of IndexExpressions: given a (list of) IndexExpr, provide
  // the (list of) Shape/Value/OpFoldResult corresponding to the original (list
  // of) IndexExpr.
  static void getShape(llvm::SmallVectorImpl<IndexExpr> &indexExprList,
      llvm::SmallVectorImpl<int64_t> &intDimList,
      bool uniqueQuestionMark = false);
  static void getValues(mlir::ArrayRef<IndexExpr> indexExprArray,
      llvm::SmallVectorImpl<mlir::Value> &valueList);
...
};
```
as well as conversion to index expression:
```C++
// Create a list of IndexExpr of kind LiteralIndexExpr from a list of integers.
void getIndexExprListFromInt(llvm::SmallVectorImpl<int64_t> &inputList,
    llvm::SmallVectorImpl<IndexExpr> &outputList);

// Create a list of IndexExpr of kind LiteralIndexExpr/Questionmark from a
// shape. Negative values are translated to Questionmarks.
void getIndexExprListFromShape(llvm::SmallVectorImpl<int64_t> &inputList,
    llvm::SmallVectorImpl<IndexExpr> &outputList);
```
and more. All in IndexExpr.hpp.

As I will suggest below, I would add this new functionality directly in IndexExpr as it will be useful to others as well. (Note that it will assert if an index expression in the list is not a literal. There are already tests to detect if they are all literal, if you wanted to give a friendlier warning.

Reply:
I will add the function to `IndexExpr`. It seems to be a really good fit with the other helper utilities!